osbuild/sources.py

```diff
@@ -3,12 +3,51 @@
 import os
 import json
 import tempfile
+from typing import List, Dict
 
+from osbuild.formats_common import OSbuildError
 from . import host
 from .objectstore import ObjectStore
 from .util.types import PathLike
 
 
+class SourceResult():
+
+    def __init__(self, source=None):
+        self.source = source
+        self.error = None
+
+    def to_dict(self) -> Dict:
+        return {
+            "source": self.source,
+            "error": self.get_error_message()
+        }
+
+    def get_error_message(self):
+        return self.error
+
+    @staticmethod
+    def from_dict(obj: Dict):
+        ret = SourceResult()
+        ret.source = obj["source"]
+        ret.error = obj.get("error", None)
+        return ret
+
+    def explode_if_error(self):
+        if self.error:
+            raise SourceException(self)
+
+
+class SourceException(OSbuildError):
```

Comment:
For the structure of `SourceError` we need to decide if we want to ever report multiple items; in that case we would need to have an array that contains the item (url, checksum) and the status code and possible description for it. If we ever only want to report one and fail fast, then we can have the fields directly at the `SourceError` level.
The same decision is true for manifest validation. Currently we don't use exceptions there and report all of them, i.e. we have a `ValidationResult` which contains individual `ValidationError`s. We probably want to change that so that  `ValidationError` is the aggregate one and then we have individual `ValidationIssue`s or something

Reply:
We can also instantiate exceptions and not raising them immediately.